src/chat/chatReducers.js

```diff
@@ -70,6 +72,29 @@ const eventMessageDelete = (
   return stateChange ? newState : state;
 };
 
+const eventUpdateMessageFlags = (
+  state: NarrowsState,
+  action: EventUpdateMessageFlagsAction,
+): NarrowsState => {
+  const { messages, flag, operation } = action;
+  const sources = [state];
+
+  if (state[STARRED_NARROW_STR] && flag === 'starred') {
+    if (operation === 'add') {
+      sources.push({
+        [STARRED_NARROW_STR]: [...state[STARRED_NARROW_STR], ...messages].sort(),
+      });
+    } else {
+      // operation === 'remove'
+      sources.push({
+        [STARRED_NARROW_STR]: state[STARRED_NARROW_STR].filter(id => !messages.includes(id)),
+      });
+    }
+  }
+
+  return Object.assign({}, ...sources);
```

Comment:
This will return a non-identical `state` even when no update happened -- e.g. when we just learned the user read some messages, which is very common.

I think this is another case where just systematically following the architecture's preferred pattern, avoiding updates unless there really is an update, doesn't appreciably complicate the code and saves complexity overall in understanding how things work.

Specifically, how about `const updates = [];`, and then `return Object.assign({}, state, ...updates);`... with an `if (!updates.length) { return state; }` just before it.

Reply:
Updated.